### PR TITLE
feat: universal tensor type

### LINKED ISSUE #782: Type: Catch-all tensor
The current implementation of `Tensor` only works on a `np.ndarray`, and should thus be renamed accordingly.

We need a `Tensor` type that can accept a tensor from "any" framework; for now that will be numpy and pytorch.

### PR DESCRIPTION
**Pull request by JohannesMessner**
Introduces the type `Tensor` can take any kind of tensor (for now this means np and torch).
The former tensor type is renamed to `NdArray`.

As far as the proto is concerned, `Tensor` does not exist; it is only a shorthand that will cast the input either to `NdArray` or to `TorchTensor`.

closes #782

### CODE REVIEW

File: docarray/predefined_document/image.py [hunk 1]
```diff
@@ -64,5 +64,5 @@ class MultiModalDoc(Document):
     """
 
     url: Optional[ImageUrl]
-    tensor: Optional[Tensor]
+    tensor: Optional[NdArray]
```
Comment (samsja) [inline docarray/predefined_document/image.py:67]: should be Tensor here

File: tests/integrations/array/test_array_proto.py [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
-from docarray.typing import Tensor
+from docarray import Document, DocumentArray, Image, Text
+from docarray.typing import NdArray
 
 
 def test_simple_proto():
     class CustomDoc(Document):
         text: str
-        tensor: Tensor
+        tensor: NdArray
```
Comment (samsja) [inline tests/integrations/array/test_array_proto.py:10]: lets keep it Tensor here

File: tests/integrations/document/test_to_json.py [hunk 3]
```diff
@@ -2,12 +2,12 @@
...
-from docarray.typing import AnyUrl, Tensor, TorchTensor
+from docarray.typing import AnyUrl, NdArray, TorchTensor
 
 
 def test_to_json():
     class Mmdoc(BaseDocument):
-        img: Tensor
+        img: NdArray
         url: AnyUrl
         txt: str
```
Comment (samsja) [inline tests/integrations/document/test_to_json.py:12]: can we add a Tensor field here ? To see if this Union does not break the to json

File: tests/integrations/externals/test_fastapi.py [hunk 4]
```diff
@@ -40,8 +40,8 @@ class InputDoc(Document):
         img: Image
 
     class OutputDoc(Document):
-        embedding_clip: Tensor
-        embedding_bert: Tensor
+        embedding_clip: NdArray
+        embedding_bert: NdArray
```
Comment (samsja) [inline tests/integrations/externals/test_fastapi.py:44]: lets keep Tensor maybe to see how fastapi is going to react

File: tests/integrations/typing/test_typing_proto.py [hunk 5]
```diff
@@ -3,12 +3,12 @@
...
 from docarray import Document
 from docarray.document import AnyDocument
-from docarray.typing import AnyUrl, Embedding, ImageUrl, Tensor, TextUrl, TorchTensor
+from docarray.typing import AnyUrl, Embedding, ImageUrl, NdArray, TextUrl, TorchTensor
 
 
 def test_proto_all_types():
     class Mymmdoc(Document):
-        tensor: Tensor
+        tensor: NdArray
```
Comment (samsja) [inline tests/integrations/typing/test_typing_proto.py:11]: keep Tensor

File: tests/units/array/test_mixins/test_attribute.py [hunk 6]
```diff
@@ -2,13 +2,13 @@
...
 from docarray.document import BaseDocument
-from docarray.typing import Tensor
+from docarray.typing import NdArray
 
 
 def test_get_bulk_attributes_function():
     class Mmdoc(BaseDocument):
         text: str
-        tensor: Tensor
+        tensor: NdArray
```
Comment (samsja) [inline tests/units/array/test_mixins/test_attribute.py:11]: keep Tensor

File: tests/units/array/test_mixins/test_attribute.py [hunk 7]
```diff
@@ -32,7 +32,7 @@ class Mmdoc(BaseDocument):
 def test_get_bulk_attributes():
     class Mmdoc(BaseDocument):
         text: str
-        tensor: Tensor
+        tensor: NdArray
```
Comment (samsja) [inline tests/units/array/test_mixins/test_attribute.py:35]: Keep Tensor

File: tests/units/document/proto/test_document_proto.py [hunk 8]
```diff
@@ -19,7 +19,7 @@ class CustomDoc(BaseDocument):
 
 def test_proto_ndarray():
     class CustomDoc(BaseDocument):
-        tensor: Tensor
+        tensor: NdArray
```
Comment (samsja) [inline tests/units/document/proto/test_document_proto.py:22]: Tensor

File: tests/units/document/proto/test_proto_based_object.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import numpy as np
 
 from docarray.proto import DocumentProto, NdArrayProto, NodeProto
-from docarray.typing import Tensor
+from docarray.typing import NdArray
```
Comment (samsja) [inline tests/units/document/proto/test_proto_based_object.py:4]: Tensor

File: tests/units/document/test_any_document.py [hunk 10]
```diff
@@ -1,13 +1,13 @@
...
 from docarray.document import AnyDocument, BaseDocument
-from docarray.typing import Tensor
+from docarray.typing import NdArray
 
 
 def test_any_doc():
     class InnerDocument(BaseDocument):
         text: str
-        tensor: Tensor
+        tensor: NdArray
```
Comment (samsja) [inline tests/units/document/test_any_document.py:10]: Tensor

File: README.md [hunk 11]
```diff
@@ -36,15 +36,14 @@ This follow [Pydantic Model](https://pydantic-docs.helpmanual.io/usage/models/)
 
 It is similar to the dataclass from the (old) docarray
 
-
 ```python
-from docarray.typing import Tensor
+from docarray.typing import NdArray
```
Comment (samsja) [inline README.md:40]: Keep Tensor as well

**Comment by JohannesMessner** [comment]
I kept the type as `NdArray` in the tests since the tests are not about that.
To properly test the (de)serialization of this new type I instead added a dedicated test for that

**Review by samsja — APPROVED** [review]

**Pull request merged**